services/headless-lms/chatbot/src/chatbot_tools/course_progress.rs

```diff
@@ -0,0 +1,172 @@
+use headless_lms_models::{ModelError, user_exercise_states::UserCourseProgress};
+use sqlx::PgConnection;
+
+use crate::{
+    azure_chatbot::ChatbotUserContext,
+    chatbot_tools::{ChatbotTool, ToolProperties},
+};
+
+type CourseProgressTool = ToolProperties<CourseProgressState, CourseProgressArguments>;
+
+impl ChatbotTool for CourseProgressTool {
+    type State = CourseProgressState;
+    type Arguments = CourseProgressArguments;
+
+    fn parse_arguments(_args_string: String) -> Self::Arguments {
+        CourseProgressArguments {}
+    }
+
+    /// Create a CourseProgressTool instance
+    async fn from_db_and_arguments(
+        conn: &mut PgConnection,
+        arguments: Self::Arguments,
+        user_context: &ChatbotUserContext,
+    ) -> Result<Self, ModelError> {
+        let progress = headless_lms_models::user_exercise_states::get_user_course_progress(
+            conn,
+            user_context.course_id,
+            user_context.user_id,
+        )
+        .await?;
+        Result::Ok(CourseProgressTool {
+            state: CourseProgressState {
+                course_name: user_context.course_name.clone(),
+                progress,
+            },
+            arguments,
+        })
+    }
+
+    /// Return a string explaining the user's progress on the course that the chatbot is on
+    fn output(&self) -> String {
+        let mut progress = self.state.progress.to_owned();
+        let course_name = &self.state.course_name;
+        let mut res = format!("The user is completing a course called {course_name}. ");
+
+        if progress.len() == 1 {
+            let module = &progress[0];
+            res.push_str("Their progress on this course is the following: ");
+
+            res = push_exercises_scores_progress(
+                res,
+                module.attempted_exercises,
+                module.total_exercises,
+                module.attempted_exercises_required,
+                module.score_given,
+                module.score_maximum,
+                module.score_required,
+            );
+        } else {
```

Comment:
Explain that the user needs to complete the base module before they can complete the additional modules.